File: src/data/orbits.ts
// ---------------------------------------------------------------------------
// orbits.ts — a small, self-contained orbital propagator for the near-Earth
// view. No npm dependency (the app ships "no runtime dependency beyond React"),
// so this is a hand-written Keplerian propagation of the TLE mean elements with
// the two dominant J2 secular drifts (nodal regression, apsidal precession,
// mean-anomaly correction) applied.
//
// What is honest here: every DISPLAYED number — altitude, orbital speed, period
// and light-time — derives from the mean motion via Kepler's third law and
// vis-viva, so it is accurate to well within a light-millisecond. What is
// approximate is the exact on-screen ANGLE (we omit SGP4's short-period and
// drag terms); no numeric value is shown for it, and today's epoch keeps the
// along-track drift small. See About copy.
// ---------------------------------------------------------------------------
import type { SatelliteRecord, OrbitBand } from '../types.ts';

export const MU = 398600.4418; // Earth GM, km^3/s^2
export const R_EARTH = 6378.137; // equatorial radius, km
const J2 = 1.08262668e-3;
const C_KM_S = 299792.458; // speed of light, km/s
const DEG = Math.PI / 180;
const TWO_PI = Math.PI * 2;
const DAY_S = 86400;

/** Semi-major axis (km) from mean motion in rev/day. */
export function semiMajorAxis(meanMotionRevPerDay: number): number {
  const n = (meanMotionRevPerDay * TWO_PI) / DAY_S; // rad/s
  return Math.cbrt(MU / (n * n));
}

/** Coarse altitude band from mean motion, used at fetch time. */
export function bandOf(meanMotionRevPerDay: number, eccentricity: number): OrbitBand {
  const a = semiMajorAxis(meanMotionRevPerDay);
  const perigee = a * (1 - eccentricity) - R_EARTH;
  const apogee = a * (1 + eccentricity) - R_EARTH;
  if (apogee - perigee > 20000) return 'HEO'; // highly eccentric (Molniya etc.)
  const alt = a - R_EARTH;
  if (alt < 2000) return 'LEO';
  if (alt < 30000) return 'MEO';
  return 'GEO';
}

export interface OrbitState {
  /** ECI position, km. */
  x: number;
  y: number;
  z: number;
  /** Geocentric distance, km. */
  r: number;
  /** Altitude above the mean surface, km. */
  altitude: number;
  /** Instantaneous orbital speed, km/s (vis-viva). */
  speed: number;
  /** Orbital period, minutes. */
  periodMin: number;
  /** One-way light time from the surface directly below (altitude / c), seconds. */
  lightSeconds: number;
}

/** Solve Kepler's equation M = E - e sinE for E (radians), Newton-Raphson. */
function eccentricAnomaly(M: number, e: number): number {
  let E = e < 0.8 ? M : Math.PI;
  for (let i = 0; i < 8; i++) {
    const dE = (E - e * Math.sin(E) - M) / (1 - e * Math.cos(E));
    E -= dE;
    if (Math.abs(dE) < 1e-10) break;
  }
  return E;
}

/**
 * Propagate one satellite's mean elements to time `atMs`.
 * Keplerian orbit + J2 secular rates on Ω, ω and M. Returns geocentric ECI
 * position and the derived scalar quantities we display.
 */
export function propagate(sat: SatelliteRecord, atMs: number): OrbitState {
  const n0 = (sat.meanMotion * TWO_PI) / DAY_S; // rad/s
  const a = Math.cbrt(MU / (n0 * n0)); // km
  const e = sat.eccentricity;
  const i = sat.inclination * DEG;

  const dt = (atMs - sat.epochMs) / 1000; // seconds since epoch

  // J2 secular rates (rad/s). p is the semi-latus rectum.
  const p = a * (1 - e * e);
  const cosi = Math.cos(i);
  const sini2 = Math.sin(i) * Math.sin(i);
  const factor = 1.5 * J2 * (R_EARTH / p) * (R_EARTH / p) * n0;
  const raanDot = -factor * cosi;
  const argpDot = factor * (2 - 2.5 * sini2);
  const mDot = n0 + factor * Math.sqrt(1 - e * e) * (1 - 1.5 * sini2);

  const raan = sat.raan * DEG + raanDot * dt;
  const argp = sat.argPerigee * DEG + argpDot * dt;
  const M = sat.meanAnomaly * DEG + mDot * dt;

  const E = eccentricAnomaly(((M % TWO_PI) + TWO_PI) % TWO_PI, e);
  const cosE = Math.cos(E);
  const sinE = Math.sin(E);
  const r = a * (1 - e * cosE); // km

  // Position in the perifocal frame.
  const xp = a * (cosE - e);
  const yp = a * Math.sqrt(1 - e * e) * sinE;

  // Rotate perifocal → ECI by (argp, inclination, raan).
  const cosO = Math.cos(raan);
  const sinO = Math.sin(raan);
  const cosw = Math.cos(argp);
  const sinw = Math.sin(argp);
  const cosI = Math.cos(i);
  const sinI = Math.sin(i);

  const x11 = cosO * cosw - sinO * sinw * cosI;
  const x12 = -cosO * sinw - sinO * cosw * cosI;
  const x21 = sinO * cosw + cosO * sinw * cosI;
  const x22 = -sinO * sinw + cosO * cosw * cosI;
  const x31 = sinw * sinI;
  const x32 = cosw * sinI;

  const x = x11 * xp + x12 * yp;
  const y = x21 * xp + x22 * yp;
  const z = x31 * xp + x32 * yp;

  const altitude = r - R_EARTH;
  const speed = Math.sqrt(MU * (2 / r - 1 / a)); // km/s
  const periodMin = (TWO_PI / n0) / 60;
  const lightSeconds = Math.max(0, altitude) / C_KM_S;

  return { x, y, z, r, altitude, speed, periodMin, lightSeconds };
}

/** Perigee altitude (km) — the decay indicator. */
export function perigeeAltitude(sat: SatelliteRecord): number {
  const a = semiMajorAxis(sat.meanMotion);
  return a * (1 - sat.eccentricity) - R_EARTH;
}

File: ios/Sublight/Orbits.swift
import Foundation

// Swift port of src/data/orbits.ts — a self-contained Keplerian + J2 propagator
// for the Near-Earth view. Displayed numbers (altitude, speed, period, light-
// time) derive from mean motion and are honest; only the on-screen angle is
// approximate (no SGP4 short-period / drag terms). Kept in lockstep with the
// web version, which is unit-tested against the ISS.

enum Orbits {
    static let mu: Double = 398_600.4418      // Earth GM, km^3/s^2
    static let rEarth: Double = 6378.137      // equatorial radius, km
    static let j2: Double = 1.08262668e-3
    static let cKmS: Double = 299_792.458     // speed of light, km/s
    private static let deg = Double.pi / 180
    private static let twoPi = Double.pi * 2
    private static let dayS: Double = 86_400

    struct State {
        var x: Double
        var y: Double
        var z: Double
        var r: Double            // geocentric distance, km
        var altitude: Double     // km above the mean surface
        var speed: Double        // km/s
        var periodMin: Double
        var lightSeconds: Double // altitude / c
    }

    static func semiMajorAxis(_ meanMotionRevPerDay: Double) -> Double {
        let n = (meanMotionRevPerDay * twoPi) / dayS   // rad/s
        return cbrt(mu / (n * n))
    }

    static func band(meanMotion: Double, eccentricity e: Double) -> String {
        let a = semiMajorAxis(meanMotion)
        let perigee = a * (1 - e) - rEarth
        let apogee = a * (1 + e) - rEarth
        if apogee - perigee > 20_000 { return "HEO" }
        let alt = a - rEarth
        if alt < 2000 { return "LEO" }
        if alt < 30_000 { return "MEO" }
        return "GEO"
    }

    static func perigeeAltitude(_ sat: SatelliteRecord) -> Double {
        let a = semiMajorAxis(sat.meanMotion)
        return a * (1 - sat.eccentricity) - rEarth
    }

    private static func eccentricAnomaly(_ M: Double, _ e: Double) -> Double {
        var E = e < 0.8 ? M : Double.pi
        for _ in 0..<8 {
            let dE = (E - e * sin(E) - M) / (1 - e * cos(E))
            E -= dE
            if abs(dE) < 1e-10 { break }
        }
        return E
    }

    /// Propagate one satellite's mean elements to `atMs` (ms since Unix epoch).
    static func propagate(_ sat: SatelliteRecord, atMs: Double) -> State {
        let n0 = (sat.meanMotion * twoPi) / dayS       // rad/s
        let a = cbrt(mu / (n0 * n0))                    // km
        let e = sat.eccentricity
        let i = sat.inclination * deg
        let dt = (atMs - sat.epochMs) / 1000            // seconds since epoch

        let p = a * (1 - e * e)
        let cosi = cos(i)
        let sini2 = sin(i) * sin(i)
        let factor = 1.5 * j2 * (rEarth / p) * (rEarth / p) * n0
        let raanDot = -factor * cosi
        let argpDot = factor * (2 - 2.5 * sini2)
        let mDot = n0 + factor * (1 - e * e).squareRoot() * (1 - 1.5 * sini2)

        let raan = sat.raan * deg + raanDot * dt
        let argp = sat.argPerigee * deg + argpDot * dt
        var M = sat.meanAnomaly * deg + mDot * dt
        M = M.truncatingRemainder(dividingBy: twoPi)
        if M < 0 { M += twoPi }

        let E = eccentricAnomaly(M, e)
        let cosE = cos(E), sinE = sin(E)
        let r = a * (1 - e * cosE)

        let xp = a * (cosE - e)
        let yp = a * (1 - e * e).squareRoot() * sinE

        let cosO = cos(raan), sinO = sin(raan)
        let cosw = cos(argp), sinw = sin(argp)
        let cosI = cos(i), sinI = sin(i)

        let x = (cosO * cosw - sinO * sinw * cosI) * xp + (-cosO * sinw - sinO * cosw * cosI) * yp
        let y = (sinO * cosw + cosO * sinw * cosI) * xp + (-sinO * sinw + cosO * cosw * cosI) * yp
        let z = (sinw * sinI) * xp + (cosw * sinI) * yp

        let altitude = r - rEarth
        let speed = (mu * (2 / r - 1 / a)).squareRoot()
        let periodMin = (twoPi / n0) / 60
        let lightSeconds = max(0, altitude) / cKmS

        return State(x: x, y: y, z: z, r: r, altitude: altitude,
                     speed: speed, periodMin: periodMin, lightSeconds: lightSeconds)
    }
}
